test/unit/codegen/codegen_helper.cpp

```diff
@@ -0,0 +1,130 @@
+
+    /// symbols/variables are collected in info object
+    const auto& info = v.analyze(*ast);
+
+    /// semicolon separated list of variables
+    std::string variables;
+
+    /// range variables in order of code generation
+    for (const auto& var: info.range_parameter_vars) {
+        variables += var->get_name() + ";";
+    }
+    for (const auto& var: info.range_assigned_vars) {
+        variables += var->get_name() + ";";
+    }
+    for (const auto& var: info.range_state_vars) {
+        variables += var->get_name() + ";";
+    }
+    for (const auto& var: info.assigned_vars) {
+        variables += var->get_name() + ";";
+    }
+
+    return variables;
+}
+
+SCENARIO("unusual / failing mod files", "[codegen][var_order]") {
+    GIVEN("cal_mig.mod : USEION variables declared as RANGE") {
+        std::string nmodl_text = R"(
+            PARAMETER {
+              gcalbar=.003 (mho/cm2)
+              ki=.001 (mM)
+              cai = 50.e-6 (mM)
+              cao = 2 (mM)
+              q10 = 5
+              USEGHK=1
+            }
+            NEURON {
+              SUFFIX cal
+              USEION ca READ cai,cao WRITE ica
+              RANGE gcalbar, cai, ica, gcal, ggk
+              RANGE minf, tau
+              GLOBAL USEGHK
+            }
+            STATE {
+              m
+            }
+            ASSIGNED {
+              ica (mA/cm2)
+              gcal (mho/cm2)
+              minf
+              tau   (ms)
+              ggk
+            }
+            DERIVATIVE state {
+              rate(v)
+              m' = (minf - m)/tau
+            }
+        )";
+
+        THEN("ionic current variable declared as RANGE appears first") {
+            std::string expected = "gcalbar;ica;gcal;minf;tau;ggk;m;cai;cao;";
+            auto result = run_inline_visitor(nmodl_text);
+            REQUIRE(result == expected);
+        }
+    }
+
+    GIVEN("CaDynamics_E2.mod : USEION variables declared as STATE variable") {
+        std::string nmodl_text = R"(
+            NEURON  {
+              SUFFIX CaDynamics_E2
+              USEION ca READ ica WRITE cai
+              RANGE decay, gamma, minCai, depth
+            }
+
+            PARAMETER   {
+              gamma = 0.05 : percent of free calcium (not buffered)
+              decay = 80 (ms) : rate of removal of calcium
+              depth = 0.1 (um) : depth of shell
+              minCai = 1e-4 (mM)
+            }
+
+            ASSIGNED {ica (mA/cm2)}
+
+            STATE {
+              cai (mM)
+            }
+
+            DERIVATIVE states   {
+              cai' = -(10000)*(ica*gamma/(2*FARADAY*depth)) - (cai - minCai)/decay
+            }
+        )";
+
+        THEN("ion state variable is ordered after parameter and assigned ionic current") {
+            std::string expected = "gamma;decay;depth;minCai;ica;cai;";
+            auto result = run_inline_visitor(nmodl_text);
+            REQUIRE(result == expected);
+        }
+    }
```

Comment:
another edge-case that could be useful to test is `thalamus/mod/ican.mod`. In there there is an ion called `n` and its current could give some problems to the ion current detection system and ispc (since `in` is a reserved variable in there)

Reply:
thanks, I will add that shortly.

Reply:
ican.mod is handled normally in the context of this PR : 

![image](https://user-images.githubusercontent.com/666852/101649610-619b9600-3a3b-11eb-9cf7-578c23988b81.png)

This mod file has issue due to `in` variable which is keyword in ISPC. This will be handled separately.